lib/src/front_end/piece_factory.dart

```diff
@@ -1604,6 +1625,14 @@ mixin PieceFactory {
     });
   }
 
+  /// Whether there is a trailing comma at the end of the list delimited by
+  /// [rightBracket].
+  bool hasPreservedTrailingComma(Token rightBracket) {
+    if (formatter.trailingCommas != TrailingCommas.preserve) return false;
```

Comment:
[nit] Consider returning the boolean expression instead of putting it in a conditional with a literal return.

```dart
    return formatter.trailingCommas == TrailingCommas.preserve &&
        rightBracket.hasCommaBefore;
 ```